anta/models.py

```diff
@@ -0,0 +1,218 @@
+"""
+Models to define a TestStructure
+"""
+from __future__ import annotations
+
+import logging
+import traceback
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from functools import wraps
+from typing import Any, Callable, ClassVar, Dict, Optional, TypeVar, Union, cast
+
+from aioeapi import EapiCommandError
+from httpx import ConnectError, HTTPError
+from pydantic import BaseModel
+
+from anta.inventory.models import InventoryDevice
+from anta.result_manager.models import TestResult
+from anta.tools import exc_to_str
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+class AntaTestCommand(BaseModel):
+    """Class to define a test command with its API version
+
+    Attributes:
+        command(str): Test command
+        version(str): eAPI version - default is latest
+        ofmt(str):  eAPI output - json or text - default is json
+        output: collected output either dict for json or str for text
+    """
+
+    command: str
+    version: str = "latest"
+    ofmt: str = "json"
+    output: Optional[Union[Dict[Any, Any], str]]
+
+
+class AntaTestFilter(ABC):
+    """Class to define a test Filter"""
+
+    # pylint: disable=too-few-public-methods
+
+    @abstractmethod
+    def should_skip(
+        self,
+        device: InventoryDevice,
+        result: TestResult,
+        *args: list[Any],
+        **kwagrs: dict[str, Any],
+    ) -> bool:
+        """
+        Sets the TestResult status to skip with the appropriate skip message
+
+        Returns:
+            bool: True if the test should be skipped, False otherwise
+        """
+
+
+class AntaTest(ABC):
+    """Abstract class defining a test for Anta
+
+    The goal of this class is to handle the heavy lifting and make
+    writing a test as simple as possible.
+
+    TODO - complete doctstring with example
+    """
+
+    # Mandatory class attributes
+    # TODO - find a way to tell mypy these are mandatory for child classes - maybe Protocol
+    name: ClassVar[str]
+    description: ClassVar[str]
+    categories: ClassVar[list[str]]
+    commands: ClassVar[list[AntaTestCommand]]
```

Comment:
need to be able to have a way for commands like `ping X` or `show ip bgp neighbor Y`